shared/src/commonMain/kotlin/org/hyperskill/app/step/presentation/StepReducer.kt

```diff
@@ -1,33 +1,71 @@
 package org.hyperskill.app.step.presentation
 
+import org.hyperskill.app.step.domain.analytic.StepViewedHyperskillAnalyticEvent
+import org.hyperskill.app.step.domain.model.StepRoute
 import org.hyperskill.app.step.presentation.StepFeature.Action
 import org.hyperskill.app.step.presentation.StepFeature.Message
 import org.hyperskill.app.step.presentation.StepFeature.State
-import org.hyperskill.app.step.domain.analytic.StepViewedHyperskillAnalyticEvent
+import org.hyperskill.app.step_completion.presentation.StepCompletionFeature
+import org.hyperskill.app.step_completion.presentation.StepCompletionReducer
 import ru.nobird.app.presentation.redux.reducer.StateReducer
 
-class StepReducer : StateReducer<State, Message, Action> {
+class StepReducer(private val stepRoute: StepRoute, private val stepCompletionReducer: StepCompletionReducer) : StateReducer<State, Message, Action> {
     override fun reduce(state: State, message: Message): Pair<State, Set<Action>> =
         when (message) {
             is Message.Initialize ->
                 if (state is State.Idle ||
                     (message.forceUpdate && (state is State.Data || state is State.Error))
                 ) {
-                    State.Loading to setOf(Action.FetchStep(message.stepRoute))
+                    State.Loading to setOf(Action.FetchStep(stepRoute))
                 } else {
                     null
                 }
-            is Message.StepLoaded.Success ->
-                State.Data(message.step) to emptySet()
+            is Message.StepLoaded.Success -> {
+                State.Data(
+                    step = message.step,
+                    isPracticingAvailable = stepRoute is StepRoute.Learn,
+                    stepCompletionState = StepCompletionFeature.createState(message.step, stepRoute)
+                ) to emptySet()
+            }
             is Message.StepLoaded.Error ->
                 State.Error to emptySet()
             is Message.ViewedEventMessage ->
                 state to setOf(
                     Action.LogAnalyticEvent(
                         StepViewedHyperskillAnalyticEvent(
-                            message.stepRoute.analyticRoute
+                            stepRoute.analyticRoute
                         )
                     )
                 )
+            is Message.StepCompletionMessage ->
+                if (state is State.Data) {
```

Comment:
`if (state is State.Data && state.isPracticingAvailable)` ?

Reply:
I don't think that we need to check isPracticingAvailable because `StepCompletionFeature` connected not only with topic completion and practicing, but potentially with daily step completion and repetition completion